### PR TITLE
feat(pn): add push notification ios native module

### PR DESCRIPTION
**Pull request by HuiSF**
#### Description of changes

Merge implementation of the react-native push notification iOS module/

#### Description of how you validated changes

Tested implementation with an example iOS App built with react-native.
Including:
* Receiving foreground notification
* Receiving background notification
* Request notification permissions
* Register device to Apple Push Notification
* Set and get App badge count
* App functionality running in background

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/rtn-push-notification/ios/AmplifyRTNPushNotification.m [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+#import <React/RCTEventEmitter.h>
+#import <React/RCTBridgeModule.h>
+
+@interface RCT_EXTERN_MODULE(AmplifyRTNPushNotification, RCTEventEmitter)
+
+RCT_EXTERN_METHOD(requestPermissions:(NSDictionary*)permissions
+                  resolve:(RCTPromiseResolveBlock)resolve
+                  reject:(RCTPromiseRejectBlock)reject)
+
```
Comment (cshfang) [inline packages/rtn-push-notification/ios/AmplifyRTNPushNotification.m:12]: Nit: Extra line break

File: packages/rtn-push-notification/ios/AmplifyRTNPushNotification.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    override func addListener(_ eventName: String!) {
+        super.addListener(eventName)
+        registeredEventNames.insert(eventName)
+
+        if (registeredEventNames == expectedEventNames) {
+            notifyFlushQueuedEvents()
+        }
+    }
+
+    override func supportedEvents() -> [String]! {
```
Comment (cshfang) [inline packages/rtn-push-notification/ios/AmplifyRTNPushNotification.swift:55]: What is this used for? Looks the same as the expected set

Comment (HuiSF) [reply packages/rtn-push-notification/ios/AmplifyRTNPushNotification.swift:55]: Oh can reuse the list here.
This method is required to be implemented extending RCTEventEmitter which ignores (and throws warnings) for a event with unexpected type.

File: packages/rtn-push-notification/ios/AmplifyRTNPushNotification.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        AmplifyRTNEventManager.shared.setSendEvent(sendEvent: self.sendEventToJS)
+    }
+
+    func sendEventToJS(event: AmplifyRTNEvent) {
+        sendEvent(withName: event.type.name, body: event.payload)
+    }
+
+    override func sendEvent(withName name: String!, body: Any!) {
+        super.sendEvent(withName: name, body: body)
+    }
```
Comment (cshfang) [inline packages/rtn-push-notification/ios/AmplifyRTNPushNotification.swift:44]: Does this need to be overridden?

**Review by cshfang — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Review by cshfang — APPROVED** [review]

**Pull request merged**